src/core/fair_queue.cc

```diff
@@ -318,37 +319,100 @@ fair_queue::clock_type::time_point fair_queue::next_pending_aio() const noexcept
+            // We can't dispatch the request yet, but we already have a pending reservation
+            // which will provide us with enough tokens for it eventually,
+            // or our reservation is already max-size and we can't reserve more tokens until we reap some.
+            // So we should just wait.
+            // We return any immediately-available tokens back to `_pending`
+            // and we bail. The next `dispatch_request` will again take those tokens
+            // (possibly joined by some newly-granted tokens) and retry.
+            _pending.cap += ready;
+            ready = 0;
             break;
-        }
-
-        if (gr == grab_result::cant_preempt) {
-            pop_priority_class(h);
-            preempt.emplace_back(&h);
+        } else if (our_turn_has_come) {
+            // The current reservation isn't enough to fulfill the next request,
+            // and we can cancel it (because `our_turn_has_come == true`) and make a bigger one
+            // (because `_pending.cap < can_grab_this_tick`).
+            // So we cancel it and do a bigger one.
+
+            // We do token recycling here: we return the tokens which we have available, and the tokens we have reserved
+            // immediately after the group head, and we return them to the bucket, immediately grabbing the same amount from the tail.
+            // This is neutral to fairness. The bandwidth we consume is still influenced only by the
+            // `max_unarmortized_reservation` portions.
+            auto recycled = ready + _pending.cap;
+            // We are grabbing more than `max_unarmortized_reservation` here, which raises a question:
+            // if each shard can actually grab more than `max_unarmortized_reservation`,
+            // then how far can the tail actually get from the head?
+            // Our fairness calculations assume that a shard can assume a fair_group queueing latency of
+            // `nr_contenders * max_unarmortized_reservation`. Was it a lie?
+            //
+            // But it can be proven that the tail doesn't get further than `max_unarmortized_reservation * nr_contenders`
+            // above the head.
+            // Proof sketch: it can be observed that at any moment each shard contributes
+            // to `tail` at most `max_unarmortized_reservation` tokens which weren't already paid by some previous addition to `head`.
+            capacity_t grab_amount = std::min<capacity_t>(recycled + max_unarmortized_reservation, _queued_capacity);
+            // There's technically nothing wrong with grabbing more than `_group.maximum_capacity()`,
+            // but the token bucket has an assert for that, and its a reasonable expectation, so let's respect that limit.
+            // It shouldn't matter in practice.
+            grab_amount = std::min<capacity_t>(grab_amount, _group.maximum_capacity());
+            _group.refund_tokens(recycled);
+            grab_capacity(grab_amount);
```

Comment:
There's no `fair_queue::grab_amount(capacity_t)` method here, only the `fair_queue::grab_capacity(const fair_queue_entry&)` one, how does it compile?

Reply:
`grab_amount` is a local variable, though? Not a function.

Reply:
`grab_amount` is a local variable, though? Not a function.

Reply:
My misprint :( updated the comment

Reply:
Huh, you're right. Apparently it works because there's an implicit conversion from `capacity_t` to `fair_queue_entry`:
https://github.com/scylladb/seastar/blob/faf6a5df9597d59b1e9d9dd60a6c6b9cc9a3608a/include/seastar/core/fair_queue.hh#L118-L119

But it does the expected thing, so... whatever, I guess?

Reply:
Please fix it [with the assumption](https://github.com/scylladb/seastar/pull/2625) that fair_queue_entry constructor is `explicit`